oneflow/user/kernels/hardtanh_kernel.cpp

```diff
@@ -0,0 +1,108 @@
+};
+
+#define REGISTER_CPU_HARDTANH_KERNEL(device, dtype)                                       \
+  REGISTER_USER_KERNEL("hardtanh")                                                        \
+      .SetCreateFn<CpuHardtanhKernel<device, dtype>>()                                    \
+      .SetIsMatchedHob((HobDeviceTag() == device)                                         \
+                       & (HobDataType("out", 0) == GetDataType<dtype>::value))            \
+      .SetInplaceProposalFn(                                                              \
+          [](const InferContext&, AddInplaceArgPair AddInplaceArgPairFn) -> Maybe<void> { \
+            OF_RETURN_IF_ERROR(AddInplaceArgPairFn("out", 0, "in", 0, true));             \
+            return Maybe<void>::Ok();                                                     \
+          });
+
+REGISTER_CPU_HARDTANH_KERNEL(DeviceType::kCPU, float)
+REGISTER_CPU_HARDTANH_KERNEL(DeviceType::kCPU, double)
+
+template<DeviceType device_type, typename T>
+class CpuHardtanhGradKernel final : public OpKernel {
+ public:
+  CpuHardtanhGradKernel() = default;
+  ~CpuHardtanhGradKernel() = default;
+
+ private:
+  void Compute(KernelComputeContext* ctx) const override {
+    const Tensor* y_tensor = ctx->Tensor4ArgNameAndIndex("y", 0);
+    const Tensor* dy_tensor = ctx->Tensor4ArgNameAndIndex("dy", 0);
+    Tensor* dx_tensor = ctx->Tensor4ArgNameAndIndex("dx", 0);
+    const T* y_ptr = y_tensor->dptr<T>();
+    const T* dy_ptr = dy_tensor->dptr<T>();
+    T* dx_ptr = dx_tensor->mut_dptr<T>();
+
+    const T min_val = static_cast<T>(ctx->Attr<double>("min_val"));
+    const T max_val = static_cast<T>(ctx->Attr<double>("max_val"));
+    const T zero_t = static_cast<T>(0);
+
+    const int32_t elem_cnt = y_tensor->shape().elem_cnt();
+    FOR_RANGE(int32_t, i, 0, elem_cnt) {
+      dx_ptr[i] = (y_ptr[i] > min_val && y_ptr[i] < max_val) ? dy_ptr[i] : zero_t;
+    }
+  }
+  bool AlwaysComputeWhenAllOutputsEmpty() const override { return false; }
+};
+
+#define REGISTER_CPU_HARDTANH_BACKWARD_KERNEL(device, dtype)                              \
+  REGISTER_USER_KERNEL("hardtanh_grad")                                                   \
+      .SetCreateFn<CpuHardtanhGradKernel<device, dtype>>()                                \
+      .SetIsMatchedHob((HobDeviceTag() == device)                                         \
+                       & (HobDataType("dx", 0) == GetDataType<dtype>::value))             \
+      .SetInplaceProposalFn(                                                              \
+          [](const InferContext&, AddInplaceArgPair AddInplaceArgPairFn) -> Maybe<void> { \
+            OF_RETURN_IF_ERROR(AddInplaceArgPairFn("dx", 0, "dy", 0, true));              \
+            return Maybe<void>::Ok();                                                     \
+          });
+
+REGISTER_CPU_HARDTANH_BACKWARD_KERNEL(DeviceType::kCPU, float)
```

Comment:
一般的习惯里，定义宏的时候是不带 `;` 的，所以使用宏的时候带上 `;`。oneflow里有时候因为宏连续使用，这个习惯并不严格遵循（有时候宏会有";"结尾）。

不过没必要刻意不遵循。
```suggestion
REGISTER_CPU_HARDTANH_BACKWARD_KERNEL(DeviceType::kCPU, float);
REGISTER_CPU_HARDTANH_BACKWARD_KERNEL(DeviceType::kCPU, double);
```